### PR TITLE
feat(gatsby-source-filesystem): Only generate hashes when a file has changed, and add an option for skipping hashing

### PR DESCRIPTION
**Pull request by FraserThompson**
## Description

Generating an MD5 hash in `gatsby-source-filesystem` is quite an expensive operation for large files, and can increase build times significantly, as well as introduce other issues with open file handles.

To mitigate this:

* Cache the inode and modification time of the file and only re-generate a hash if this has changed
* Add a "fastHash" option which skips MD5 hashing altogether, instead just using the inode and modification time as the contentDigest, which can be used on sites with many large files which still experience build time issues

### Documentation

I've added a section to the README explaining the new option. Not sure if what I wrote goes into sufficient detail (or too much detail) though.

## Related Issues

Addresses #12011
Fixes #37425

### CODE REVIEW

File: packages/gatsby-source-filesystem/src/create-file-node.js [hunk 1]
```diff
@@ -35,7 +36,21 @@ exports.createFileNode = async (
...
+    let contentDigest
+
+    if (pluginOptions.fastHash) {
+      // Skip hashing.
+      contentDigest = key
+    } else {
+      // Generate a hash, but only if the file has changed.
+      contentDigest = cache && (await cache.get(key))
+      if (!contentDigest) {
+        contentDigest = await md5File(slashedFile.absolutePath)
```
Comment (LekoArts) [inline packages/gatsby-source-filesystem/src/create-file-node.js:49]: My PR https://github.com/gatsbyjs/gatsby/pull/37433 got merged, so you can rebase onto master and use the md5File function from there now.

```js
const { md5File } = require(`gatsby-core-utils`)
```

The `md5-file` package can be uninstalled then :)

File: packages/gatsby-source-filesystem/src/__tests__/create-file-node.js [hunk 2]
```diff
@@ -59,77 +151,74 @@ describe(`create-file-node`, () => {
...
+    } finally {
+      fs.rmdirSync(dname)
+    }
+  })
+
+  it(`records the shape of the node from cache`, async () => {
+    const dname = fs.mkdtempSync(`gatsby-create-file-node-test`).trim()
+    try {
+      const fname = path.join(dname, `f`)
+      console.log(dname, fname)
```
Comment (LekoArts) [inline packages/gatsby-source-filesystem/src/__tests__/create-file-node.js:177]: Stray console.log

File: packages/gatsby-source-filesystem/src/__tests__/create-file-node.js [hunk 3]
```diff
@@ -59,77 +151,74 @@ describe(`create-file-node`, () => {
...
+    } finally {
+      fs.rmdirSync(dname)
+    }
+  })
+
+  it(`records the shape of the fast hashed node`, async () => {
+    const dname = fs.mkdtempSync(`gatsby-create-file-node-test`).trim()
+    try {
+      const fname = path.join(dname, `f`)
+      console.log(dname, fname)
```
Comment (LekoArts) [inline packages/gatsby-source-filesystem/src/__tests__/create-file-node.js:206]: Stray console.log

File: packages/gatsby-source-filesystem/src/gatsby-node.js [hunk 4]
```diff
@@ -162,6 +169,7 @@ exports.pluginOptionsSchema = ({ Joi }) =>
   Joi.object({
     name: Joi.string(),
     path: Joi.string(),
+    fastHash: Joi.boolean(),
```
Comment (LekoArts) [inline packages/gatsby-source-filesystem/src/gatsby-node.js:172]:
```suggestion
    fastHash: Joi.boolean().default(false),
```

**Review by LekoArts — CHANGES_REQUESTED** [review]
Thanks for the PR! I'll change the README a little bit more later, so code wise I only have some small requests :)

File: packages/gatsby-source-filesystem/src/__tests__/create-file-node.js (hunk 2 above)
Comment (FraserThompson) [reply packages/gatsby-source-filesystem/src/__tests__/create-file-node.js]: It was there already so I left it in but I've removed them now

**Review by LekoArts — APPROVED** [review]
Thank you! A lot of people will benefit from that ❤️

**Pull request merged**